platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ImportMigrationConfigurationAdminContext.java

```diff
@@ -0,0 +1,312 @@
+                        if (mcfgs.isEmpty()) {
+                            memoryFactoryServices.remove(fpid);
+                        }
+                        return mcfg;
+                    }
+                }
+            }  // else - this means we will not be able to correlate an exported managed service factory
+            //           with its counterpart here, as such we will be forced to treat it as a new one
+            return null;
+        }
+        return memoryServices.remove(entry.getPid());
+    }
+
+    private ImportMigrationConfigurationAdminEntry proxy(ImportMigrationEntry entry) {
+        final Path path = entry.getPath();
+        final String extn = FilenameUtils.getExtension(path.toString());
+        final PersistenceStrategy ps = admin.getPersister(extn);
+
+        if (ps == null) {
+            getReport().record(new MigrationException(
+                    "Import error: persistency strategy [%s] for configuration [%s] is not defined.",
+                    extn,
+                    path));
+        } else {
+            final Dictionary<String, Object> properties;
+            InputStream is = null;
+
+            try {
+                is = entry.getInputStream()
+                        .orElse(null);
+                if (is == null) {
+                    throw new MigrationException(
+                            "Import error: failed to read configuration [%s]; not exported.",
+                            path);
+                }
+                try {
+                    properties = ps.read(is);
+                } catch (IOException e) {
+                    throw new MigrationException(
+                            "Import error: failed to read configuration [%s] using persistent strategy [%s]; %s.",
+                            path,
+                            ps.getExtension(),
+                            e);
+                }
+            } catch (IOException e) {
+                throw new MigrationException("Import error: failed to read configuration [%s]; %s.",
+                        path,
+                        e);
+            } finally {
+                IOUtils.closeQuietly(is);
+            }
+            try {
+                return new ImportMigrationConfigurationAdminEntry(this, entry, properties);
+            } catch (MigrationException e) {
+                // don't throw it back yet as we want to detect as many as possible so just record it
+                getReport().record(e);
+            }
+        }
+        this.isValid = false;
+        return null;
+    }
+
+    @Nullable
+    private Path getPathFromConfiguration(@Nullable Dictionary<String, Object> properties,
+            Supplier<String> from) {
+        if (properties == null) {
+            return null;
+        }
+        final Object o = properties.get(DirectoryWatcher.FILENAME);
+        final Path path;
+
+        if (o != null) {
+            try {
+                if (o instanceof URL) {
```

Comment:
Reading code like this, I really wish we were in a position to bring Kotlin into DDF.

Reply:
Nothing I can do about it at the moment. This code is actually extracted from Karaf.